scapy/layers/can.py

```diff
@@ -84,6 +88,211 @@ def extract_padding(self, p):
+
+        pkt.wirelen = max(pkt.wirelen, field_len)
+
+        fld_val = struct.unpack(fmt, (s + b"\x00" * 8)[:8])[0] >> shift
+        fld_val &= ((1 << self.size) - 1)
+
+        if self._is_float_number():
+            fld_val = struct.unpack(self.fmt,
+                                    struct.pack(self.fmt[0] + "I", fld_val))[0]
+        elif self._is_signed_number():
+            fld_val = self._convert_to_signed(fld_val, self.size)
+
+        return s, self.m2i(pkt, fld_val)
+
+    def randval(self):
+        if self._is_float_number():
+            return RandBinFloat(0, 0)
+
+        if self._is_signed_number():
+            min_val = -2**(self.size - 1)
+            max_val = 2**(self.size - 1) - 1
+        else:
+            min_val = 0
+            max_val = 2 ** self.size - 1
+
+        min_val = round(min_val * self.scaling + self.offset, self.ndigits)
+        max_val = round(max_val * self.scaling + self.offset, self.ndigits)
+
+        return RandFloat(min(min_val, max_val), max(min_val, max_val))
+
+    def i2len(self, pkt, x):
+        return float(self.size) / 8
+
+
+class LEUnsignedSignalField(SignalField):
+    def __init__(self, name, default, start, size, scaling=1, unit="",
+                 offset=0, ndigits=3):
+        SignalField.__init__(self, name, default, start, size,
+                             scaling, unit, offset, ndigits, "<B")
+
+
+class LESignedSignalField(SignalField):
+    def __init__(self, name, default, start, size, scaling=1, unit="",
+                 offset=0, ndigits=3):
+        SignalField.__init__(self, name, default, start, size,
+                             scaling, unit, offset, ndigits, "<b")
+
+
+class BEUnsignedSignalField(SignalField):
+    def __init__(self, name, default, start, size, scaling=1, unit="",
+                 offset=0, ndigits=3):
+        SignalField.__init__(self, name, default, start, size,
+                             scaling, unit, offset, ndigits, ">B")
+
+
+class BESignedSignalField(SignalField):
+    def __init__(self, name, default, start, size, scaling=1, unit="",
+                 offset=0, ndigits=3):
+        SignalField.__init__(self, name, default, start, size,
+                             scaling, unit, offset, ndigits, ">b")
+
+
+class LEFloatSignalField(SignalField):
+    def __init__(self, name, default, start, size, scaling=1, unit="",
+                 offset=0, ndigits=3):
+        SignalField.__init__(self, name, default, start, size,
+                             scaling, unit, offset, ndigits, "<f")
+
+
+class BEFloatSignalField(SignalField):
+    def __init__(self, name, default, start, size, scaling=1, unit="",
+                 offset=0, ndigits=3):
+        SignalField.__init__(self, name, default, start, size,
+                             scaling, unit, offset, ndigits, ">f")
+
+
+class SignalPacket(Packet):
+    def pre_dissect(self, s):
+        if not all([isinstance(f, SignalField) for f in self.fields_desc]):
```

Comment:
A generator could be used here
```suggestion
        if not all(isinstance(f, SignalField) for f in self.fields_desc):
```

Reply:
Otherwise all good.

I wish it didn't require the `self.raw_packet_cache = None`, because I don't want it to be I'm the usual (it remains somewhat of a hack), but it seems it does.

Reply:
Done